Basic LPUART support for STM32Lx

tagunil opened the pull request:
Tested on STM32L073.

tagunil committed:
drivers: serial: stm32: enable LBD only for UARTs with LIN support

LBD interrupt manupulation makes sense for UART with LIN support only.
Otherwise this bit should not be touched.

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>

tagunil committed:
arch: arm: soc: stm32l0: add LPUART1 pinmux options

Add obvious and easily testable PA2/PA3 configuration
(same pins as USART2).

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>

erwango commented:
@vaussard , I can't add you to the review, but would be nice if you can have a look.

vaussard commented on drivers/serial/Kconfig.stm32:
```diff
@@ -7,10 +7,12 @@
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
-	depends on SOC_FAMILY_STM32
+	select USE_STM32_LL_USART
```
Selecting USE_STM32_LL_USART does not look related to LPUART. Is that a separate fix?

tagunil replied:
Yes, it's just a fixing things around. I can separate it too.

tagunil replied:
Though it should be fixed before adding LPUART.

erwango replied:
I don't think this was actually required for U(S)ART driver.
Stm32 serial driver only used LL headers until now, which do not require this flag.

vaussard commented on drivers/serial/uart_stm32.c:
```diff
@@ -287,15 +287,28 @@ static int uart_stm32_init(struct device *dev)
 #endif
-			     config->baud_rate);
+				     config->baud_rate);
+	}
```
This part starts to be really hard to read due to the excessive ifdef'ery. I had to read twice to understand why the `if (1)` in the middle. What about completely decoupling STM32Lxx from the rest as both parts have few in common?
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || defined(CONFIG_SOC_SERIES_STM32L4X)
    if (IS_LPUART_INSTANCE(UartInstance)) {
        LL_LPUART_SetBaudRate(...);
    }
#else /* CONFIG_SOC_SERIES_STM32Lxx */
    LL_USART_SetBaudRate(...);
#endif
```

tagunil replied:
Well, then heavily ifdef'ed code for ordinary USARTs will be duplicated like that:
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || defined(CONFIG_SOC_SERIES_STM32L4X)
    if (IS_LPUART_INSTANCE(UartInstance)) {
        LL_LPUART_SetBaudRate(...);
    } else {
        LL_USART_SetBaudRate(...);
    }
#else /* CONFIG_SOC_SERIES_STM32Lxx */
    LL_USART_SetBaudRate(...);
#endif
```
Is that really better?

erwango replied:
Another option is to add use a new function either in this file, either in new file uart_lx_stm32.c

tagunil replied:
I've tried to make this thing more readable in today's push. Could you look at that?

vaussard replied:
It is far more readable, thank you.

vaussard commented on dts/arm/st/stm32l4.dtsi:
```diff
@@ -90,6 +90,15 @@
+		lpuart1: serial@40008000 {
+			compatible = "st,stm32-lpuart";
+			reg = <0x40008000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00004000>;
```
It should be `clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000001>;` as bit 0 is used on L4.

tagunil replied:
Thanks, I'll look at that!

tagunil replied:
Yes, it's just stupid copy-paste mistakes. I'll fix that.

vaussard commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -80,6 +80,15 @@
 		};
 
+		lpuart1: serial@40004800 {
+			compatible = "st,stm32-lpuart";
```
The compatible should probably be `"st,stm32-lpuart", "st,stm32-uart"` as we are using the common driver for all UART flavours. I checked the Linux binding for STM32, but it does not have support for LPUART.

tagunil replied:
I'll fix that.

vaussard reviewed (commented):
Hi @erwango, it is done. I did a similar change myself as while ago to support the LPUART, so it looks good apart a few minor comments.

vaussard commented:
Mmmh why are some of my comments shown as outdated, while the code did not changed? Strange github.

tagunil commented on drivers/serial/uart_stm32.c:
```diff
@@ -291,7 +291,7 @@ static int uart_stm32_init(struct device *dev)
 			     LL_USART_OVERSAMPLING_16,
 #endif
-			     data->huart.Init.BaudRate);
+			     config->baud_rate);
```
No, you are right, that should be done in a separate commit.

tagunil commented:
@vaussard, thank you for detailed review!

tagunil committed:
drivers: serial: stm32: remove HAL driver legacy

Remove UART_HandleTypeDef instance, which is used
only to store baudrate value.

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>

tagunil committed:
drivers: serial: stm32: rework macros and fixups

STM32Lx LPUART peripherals do not fit well into existing U(S)ART
port numbering scheme, so there will be two separate namespaces
in Kconfig: one for U(S)ARTs and one for LPUARTs.

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>

tagunil committed:
drivers: serial: stm32: add LPUART support for L0/L4 series

LPUART (Low-power UART) peripheral is just like ordinary U(S)ART
which lives in a separate clock/power domain.
Therefore already existing code could be reused as is
almost entirely.

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>

erwango commented on dts/arm/st/stm32l0-pinctrl.dtsi:
```diff
@@ -21,6 +21,12 @@
 					tx = <STM32_PIN_PA2 (STM32_PINMUX_ALT_FUNC_4 | STM32_PUSHPULL_NOPULL)>;
 				};
 			};
+			lpuart1_pins_a: lpuart1@0 {
```
Unfortunately, pin generation is not there yet, so to be actually tested, updating boards pinmux.c will be required for now.

tagunil replied:
I've done that in my working copy to test. Nucleo-L073RZ is good for such testing, because LPUART1 is muxed with USART2. I don't think however that  some board pinmux should be modified permanently.

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -7,10 +7,11 @@
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
-	depends on SOC_FAMILY_STM32
+	select USE_STM32_LL_USART
```
I don't think this was actually required for U(S)ART driver which actually compiled w/o this flag activated.
STM32 serial driver only used LL headers until now, which does not require this flag.
Please re-check and remove

tagunil replied:
I see. Headers are sufficient. Will do.

tagunil replied:
Done.

tagunil commented:
@erwango, could you look again?
